### PR TITLE
`ConfigEditor` updates

### PR DESCRIPTION
**Pull request by aangelisc**
For standardisation across Grafana we're updating datasource config editors to be more consistent. This is a part of https://github.com/grafana/oss-plugin-partnerships/issues/225.

**This is a breaking change as the plugin now has a minimum Grafana dependency of 9.5.0**

The config editor has been updated to replace `InlineField` with `Field` and the config editor has been refactored to use the new config editor structural components. There is also now individual validation on required fields. Node version has also been upgraded to 18.

image

Fixes grafana/oss-plugin-partnerships#254

### CODE REVIEW

File: src/utils/version.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import { isNumber } from 'lodash'
+
+const versionPattern = /^(\d+)(?:\.(\d+))?(?:\.(\d+))?(?:-([0-9A-Za-z\.]+))?/
+
+export class SemVersion {
```
Comment (adamyeats) [inline src/utils/version.ts:5]: Could we not just use the `semver` package for this?

File: .config/webpack/webpack.config.ts [hunk 2]
```diff
@@ -13,189 +13,201 @@ import path from 'path';
...
+                  decorators: false,
+                  dynamicImport: true,
+                },
               },
             },
           },
         },
+        {
+          test: /\.css$/,
+          use: ["style-loader", "css-loader"]
```
Comment (adamyeats) [inline .config/webpack/webpack.config.ts:100]: These quotes don't seem to be consistent with the rest of the file.

File: src/components/Divider.tsx [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
+import { Divider as GrafanaDivider, useTheme2 } from '@grafana/ui';
+import { config } from '@grafana/runtime';
+import { isVersionGtOrEq } from '../utils/version';
+
+export function Divider() {
+  const theme = useTheme2();
+  return isVersionGtOrEq(config.buildInfo.version, '10.1.0') ? (
```
Comment (adamyeats) [inline src/components/Divider.tsx:8]: I think this reads better as just a regular `if`/`else` that returns early if `true`.

File: .config/webpack/webpack.config.ts [hunk 4]
```diff
@@ -13,189 +13,201 @@ import path from 'path';
...
-                dynamicImport: true,
+
+    context: path.join(process.cwd(), SOURCE_DIR),
+
+    devtool: env.production ? 'source-map' : 'eval-source-map',
+
+    entry: await getEntries(),
+
+    externals: [
+      'lodash',
```
Comment (adamyeats) [inline .config/webpack/webpack.config.ts:37]: Should we alphabetise these array values?

File: src/editors/SentryConfigEditor.tsx [hunk 5]
```diff
@@ -38,64 +42,80 @@ export const SentryConfigEditor = (props: SentryConfigEditorProps) => {
...
-              className="reset-button"
-              onClick={() => {
-                setAuthToken('');
-                onSecureOptionChange('authToken', authToken, false);
-              }}
-            >
-              {ConfigEditorSelectors.SentrySettings.AuthToken.Reset.label}
-            </Button>
-          </>
+          <Field
```
Comment (adamyeats) [inline src/editors/SentryConfigEditor.tsx:90]: There's a lot going on in this `return` statement, is there a possibility to break some of the inline components that render conditionally into new discrete components?

File: package.json [hunk 6]
```diff
@@ -10,11 +10,11 @@
     "build": "webpack -c ./.config/webpack/webpack.config.ts --env production",
     "dev": "webpack -w -c ./.config/webpack/webpack.config.ts --env development",
     "dev_mock": "docker-compose -f docker-compose-mock.yml up",
-    "e2e": "yarn cypress install && yarn grafana-e2e run",
+    "e2e": "yarn exec cypress install && yarn exec grafana-e2e run",
```
Comment (adamyeats) [inline package.json:13]: Nice, like that you've considered `yarn >= 2` users here.

**Review by adamyeats — COMMENTED** [review]
👍 Looks good! I'd be keen to see if the `semver` package can be used instead of having our own code for this, it's so commonly used that it's probably already in our deps tree somewhere so we're probably OK just to depend on it ourselves.

File: src/utils/version.ts (hunk 1 above)
Comment (aangelisc) [reply src/utils/version.ts:5]: Iirc the `semver` package does not handle [pre-release versions well](https://github.com/npm/node-semver#prerelease-tags). We could double check but I think we've ran into these kinds of issues before. I'm not a big fan of the duplicated code so if we can get this working it'll be nicer.

File: .config/webpack/webpack.config.ts (hunk 2 above)
Comment (aangelisc) [reply .config/webpack/webpack.config.ts:100]: Anything in the `.config` directory is autogenerated so I've left it as is and we can PR the upstream library if needed

File: .config/webpack/webpack.config.ts (hunk 4 above)
Comment (aangelisc) [reply .config/webpack/webpack.config.ts:37]: Also autogenerated 😊

File: src/editors/SentryConfigEditor.tsx (hunk 5 above)
Comment (aangelisc) [reply src/editors/SentryConfigEditor.tsx:90]: We could potentially break out the secure socks component...

**Review by adamyeats — APPROVED** [review]

**Pull request merged**